network/peers/conn_manager.go

```diff
@@ -64,48 +72,122 @@ func (c connManager) TrimPeers(ctx context.Context, logger *zap.Logger, net libp
 	// c.connManager.TrimOpenConns(ctx)
 	for _, pid := range allPeers {
 		if !c.connManager.IsProtected(pid, protectedTag) {
-			_ = net.ClosePeer(pid)
-			// err := net.ClosePeer(pid)
+			err := net.ClosePeer(pid)
+			logger.Debug("DUMP: closing peer", zap.String("pid", pid.String()), zap.Error(err))
 			// if err != nil {
 			//	logger.Debug("could not close trimmed peer",
 			//		zap.String("pid", pid.String()), zap.Error(err))
 			//}
 		}
 	}
-	logger.Debug("after trimming of peers", zap.Int("beforeTrim", before),
+	logger.Debug("trimmed peers", zap.Int("beforeTrim", before),
 		zap.Int("afterTrim", len(net.Peers())))
 }
 
 // getBestPeers loop over all the existing peers and returns the best set
 // according to the number of shared subnets,
 // while considering subnets with low peer count to be more important.
-func (c connManager) getBestPeers(n int, mySubnets records.Subnets, allPeers []peer.ID, topicMaxPeers int) map[peer.ID]int {
-	peerScores := make(map[peer.ID]int)
+func (c connManager) getBestPeers(n int, mySubnets records.Subnets, allPeers []peer.ID, topicMaxPeers int) map[peer.ID]PeerScore {
+	peerScores := make(map[peer.ID]PeerScore)
 	if len(allPeers) < n {
 		for _, p := range allPeers {
 			peerScores[p] = 1
 		}
 		return peerScores
 	}
 	stats := c.subnetsIdx.GetSubnetsStats()
-	minSubnetPeers := 2
+	minSubnetPeers := 4
 	subnetsScores := GetSubnetsDistributionScores(stats, minSubnetPeers, mySubnets, topicMaxPeers)
+
+	var peerLogs []peerLog
 	for _, pid := range allPeers {
-		var peerScore int
-		subnets := c.subnetsIdx.GetPeerSubnets(pid)
-		for subnet, val := range subnets {
-			if val == byte(0) && subnetsScores[subnet] < 0 {
-				peerScore -= subnetsScores[subnet]
-			} else {
-				peerScore += subnetsScores[subnet]
-			}
+		peerSubnets := c.subnetsIdx.GetPeerSubnets(pid)
+		var score PeerScore
+		if len(peerSubnets) == 0 {
+			// TODO: shouldn't we not connect to peers with no subnets?
+			c.logger.Debug("peer has no subnets", zap.String("peer", pid.String()))
+			score = -1000
+		} else {
+			score = scorePeer(peerSubnets, subnetsScores)
 		}
-		// adding the number of shared subnets to the score, considering only up to 25% subnets
-		shared := records.SharedSubnets(subnets, mySubnets, len(mySubnets)/4)
-		peerScore += len(shared) / 2
-		// logger.Debug("peer score", zap.String("id", pid.String()), zap.Int("score", peerScore))
-		peerScores[pid] = peerScore
+		peerScores[pid] = score
+		peerLogs = append(peerLogs, peerLog{
+			Peer:          pid,
+			Score:         score,
+			SharedSubnets: len(records.SharedSubnets(peerSubnets, mySubnets, len(mySubnets))),
+		})
 	}
 
+	c.logPeerScores(peerLogs, mySubnets, stats.Connected)
+
 	return GetTopScores(peerScores, n)
 }
+
+type peerLog struct {
+	Peer          peer.ID
+	Score         PeerScore
+	SharedSubnets int
+}
+
+func (c connManager) logPeerScores(peerLogs []peerLog, mySubnets records.Subnets, subnetConnections []int) {
```

Comment:
this code is a bit complicated and thats not a good thing. Also not something we going to change right now but this peer-balancing according to subnets is an attack vector, because we report about ourselves the subnets we are in. And you can send whatever subnets you want, this affects scoring and can help an attacker eclipse our node.
it's pretty easy actually as in the handshake you send me the subnets you are in, I can understand you want these subnets so I send you back that I'm in all of them. and when balancing you will kick out peers that are not present in all of them like me (of course if I keep connecting with more peers that share all the subnets)
it's worth to research about how libp2p recommends to do peer scoring as subnets are just topics eventually.

Reply:
> this code is a bit complicated and thats not a good thing. Also not something we going to change right now but this peer-balancing according to subnets is an attack vector, because we report about ourselves the subnets we are in. And you can send whatever subnets you want, this affects scoring and can help an attacker eclipse our node. it's pretty easy actually as in the handshake you send me the subnets you are in, I can understand you want these subnets so I send you back that I'm in all of them. and when balancing you will kick out peers that are not present in all of them like me (of course if I keep connecting with more peers that share all the subnets) it's worth to research about how libp2p recommends to do peer scoring as subnets are just topics eventually.

Good call, let's create a GitHub issue to rethink subnet negotiation?